### PR TITLE
Add api endpoint to query burn fee by reserve (hourly, daily)

### PR DESCRIPTION
**Pull request by namnm1991**

### CODE REVIEW

**Comment by namnm1991** [comment]
[Burn Fee Events components](https://trello.com/c/MPfmNIMD/260-burn-fee-events-components)

File: tradelogs/http/http.go [hunk 1]
```diff
@@ -25,6 +28,29 @@ type tradeLogsQuery struct {
...
+	From        uint64 `form:"from"`
+	To          uint64 `form:"to"`
+	Freq        string `form:"freq"`
+	ReserveAddr string `form:"reserveAddr"`
+}
+
+func validateTimeWindow(fromTime, toTime time.Time, freq string) error {
+	switch strings.ToLower(freq) {
+	case "h":
+		if toTime.After(fromTime.Add(time.Hour * 24 * 180)) {
```
Comment (favadi) [inline tradelogs/http/http.go]: this should be a const.

File: tradelogs/http/http.go [hunk 2]
```diff
@@ -25,6 +28,29 @@ type tradeLogsQuery struct {
...
+			return errors.New("hourly frequency limit is 180 days")
+		}
+	case "d":
+		if toTime.After(fromTime.Add(time.Hour * 24 * 365 * 3)) {
```
Comment (favadi) [inline tradelogs/http/http.go]: this should be a const

File: tradelogs/http/http.go [hunk 3]
```diff
@@ -25,6 +28,29 @@ type tradeLogsQuery struct {
...
+	ReserveAddr string `form:"reserveAddr"`
+}
+
+func validateTimeWindow(fromTime, toTime time.Time, freq string) error {
+	switch strings.ToLower(freq) {
+	case "h":
+		if toTime.After(fromTime.Add(time.Hour * 24 * 180)) {
+			return errors.New("hourly frequency limit is 180 days")
+		}
+	case "d":
```
Comment (favadi) [inline tradelogs/http/http.go]: "h" and "d"  should be a cosnt as well.

File: tradelogs/http/http.go [hunk 4]
```diff
@@ -25,6 +28,29 @@ type tradeLogsQuery struct {
...
+	To          uint64 `form:"to"`
+	Freq        string `form:"freq"`
+	ReserveAddr string `form:"reserveAddr"`
+}
+
+func validateTimeWindow(fromTime, toTime time.Time, freq string) error {
+	switch strings.ToLower(freq) {
+	case "h":
+		if toTime.After(fromTime.Add(time.Hour * 24 * 180)) {
+			return errors.New("hourly frequency limit is 180 days")
```
Comment (favadi) [inline tradelogs/http/http.go]: the error should be `fmt.Errorf("something contains above time duration")`

File: tradelogs/http/http.go [hunk 5]
```diff
@@ -25,6 +28,29 @@ type tradeLogsQuery struct {
...
+		}
+	case "d":
+		if toTime.After(fromTime.Add(time.Hour * 24 * 365 * 3)) {
+			return errors.New("daily frequency limit is 3 years")
```
Comment (favadi) [inline tradelogs/http/http.go]: the error should be fmt.Errorf("something contains above time duration")

File: tradelogs/http/http.go [hunk 6]
```diff
@@ -25,6 +28,29 @@ type tradeLogsQuery struct {
...
+		}
+	default:
+		return errors.New("invalid frequency")
```
Comment (favadi) [inline tradelogs/http/http.go]: maybe print out supported values?

File: tradelogs/http/http.go [hunk 7]
```diff
@@ -68,9 +94,63 @@ func (ha *Server) getTradeLogs(c *gin.Context) {
...
+	var query burnFeeQuery
+	if err := c.ShouldBindQuery(&query); err != nil {
+		c.JSON(
+			http.StatusBadRequest,
+			gin.H{"error": err.Error()},
+		)
+		return
+	}
+
+	if !ethereum.IsHexAddress(query.ReserveAddr) {
```
Comment (favadi) [inline tradelogs/http/http.go]: this check could be automated by import  `_ "github.com/KyberNetwork/reserve-stats/lib/httputil` and add `binding:"required,isAddress"` to `reserveAddr` field.

File: tradelogs/http/http.go [hunk 8]
```diff
@@ -68,9 +94,63 @@ func (ha *Server) getTradeLogs(c *gin.Context) {
...
+		ha.sugar.Debug("invalid reserve address")
+		c.JSON(
+			http.StatusBadRequest,
+			gin.H{"error": "invalid reserve address"},
+		)
+		return
+	}
+
+	fromTime := time.Unix(0, int64(query.From)*int64(time.Millisecond))
+	toTime := time.Unix(0, int64(query.To)*int64(time.Millisecond))
```
Comment (favadi) [inline tradelogs/http/http.go]: we have lib/timeutil.TimestampMsToTime for this already.

File: tradelogs/http/http.go [hunk 9]
```diff
@@ -68,9 +94,63 @@ func (ha *Server) getTradeLogs(c *gin.Context) {
...
+
+	if err := validateTimeWindow(fromTime, toTime, query.Freq); err != nil {
+		c.JSON(
+			http.StatusBadRequest,
+			gin.H{"error": err.Error()},
+		)
+		return
+	}
+
+	if toTime.Equal(time.Unix(0, 0)) {
```
Comment (favadi) [inline tradelogs/http/http.go]: there is `time.IsZero` for this.

File: tradelogs/http/http.go [hunk 10]
```diff
@@ -68,9 +94,63 @@ func (ha *Server) getTradeLogs(c *gin.Context) {
...
+		toTime = time.Now()
+		fromTime = toTime.Add(-time.Hour)
```
Comment (favadi) [inline tradelogs/http/http.go]: we should only do that if fromTime is zero, right?

File: tradelogs/http/http.go [hunk 11]
```diff
@@ -25,6 +28,29 @@ type tradeLogsQuery struct {
 	To   uint64 `form:"to"`
 }
 
+type burnFeeQuery struct {
+	From        uint64 `form:"from"`
+	To          uint64 `form:"to"`
+	Freq        string `form:"freq"`
+	ReserveAddr string `form:"reserveAddr"`
```
Comment (favadi) [inline tradelogs/http/http.go]: could you doble check the same parameter is used in @cyberliem 's reserve rates API for passing reserve address? Also, the reserve rates API support passing multiple reserves addresses at the same time, could we have this feature here?

File: tradelogs/storage/influx.go [hunk 12]
```diff
@@ -67,6 +68,57 @@ func (is *InfluxStorage) SaveTradeLogs(logs []common.TradeLog, rates []tokenrate
...
+	var (
+		measurement string
+	)
+	logger := is.sugar.With("from", from, "to", to, "freq", freq, "reserveAddr", reserveAddr)
+
+	switch strings.ToLower(freq) {
+	case "h":
+		measurement = "hourly_burn_fees"
+	case "d":
+		measurement = "daily_burn_fees"
```
Comment (favadi) [inline tradelogs/storage/influx.go]: those mapping should be in var global const.

**Pull request merged**